device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceIO.java

```diff
@@ -185,24 +177,30 @@ void addClient(DeviceClientConfig config)
     /**
      * Handles logic common to all open functions.
      */
-    private void commonOpenSetup()
+    private void startWorkerThreads()
     {
         this.sendTask = new IotHubSendTask(this.transport);
         this.receiveTask = new IotHubReceiveTask(this.transport);
 
-        this.taskScheduler = Executors.newScheduledThreadPool(2);
```

Comment:
Splitting this into two separate schedulers so that there is no chance of this thread pool of size 2 having two send tasks active at once, and both waiting for work to arrive. If that were to happen, then no receive threads will spawn, so any received messages will be stuck.